src/traces/densitymapbox/plot.js

```diff
@@ -0,0 +1,131 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var convert = require('./convert');
+
+function DensityMapbox(subplot, uid) {
+    this.subplot = subplot;
+    this.uid = uid;
+
+    this.sourceId = uid + '-source';
+
+    this.layerList = [
+        ['heatmap', uid + '-layer-heatmap']
+    ];
+
+    // previous 'below' value,
+    // need this to update it properly
+    this.below = null;
+}
+
+var proto = DensityMapbox.prototype;
+
+proto.update = function(calcTrace) {
+    var subplot = this.subplot;
+    var layerList = this.layerList;
+    var optsAll = convert(calcTrace);
+
+    subplot.map
+        .getSource(this.sourceId)
+        .setData(optsAll.geojson);
+
+    if(optsAll.below !== this.below) {
+        this._removeLayers();
+        this._addLayers(optsAll);
+    }
+
+    for(var i = 0; i < layerList.length; i++) {
+        var item = layerList[i];
+        var k = item[0];
+        var id = item[1];
+        var opts = optsAll[k];
+
+        subplot.setOptions(id, 'setLayoutProperty', opts.layout);
+
+        if(opts.layout.visibility === 'visible') {
+            subplot.setOptions(id, 'setPaintProperty', opts.paint);
+        }
+    }
+};
+
+proto._addLayers = function(optsAll) {
+    var subplot = this.subplot;
+    var layerList = this.layerList;
+    var sourceId = this.sourceId;
+    var below = this.getBelow(optsAll);
+
+    for(var i = 0; i < layerList.length; i++) {
+        var item = layerList[i];
+        var k = item[0];
+        var opts = optsAll[k];
+
+        subplot.map.addLayer({
+            type: k,
+            id: item[1],
+            source: sourceId,
+            layout: opts.layout,
+            paint: opts.paint
+        }, below);
+    }
+
+    this.below = below;
+};
+
+proto._removeLayers = function() {
+    var map = this.subplot.map;
+    var layerList = this.layerList;
+
+    for(var i = 0; i < layerList.length; i++) {
```

Comment:
The loop order may be reversed here.
Removing from last item of the list towards the first item may have the benefit of having the valid index. 
On another note - I was wondering if the list should be set to an empty array at the end of this process?

Reply:
Done in https://github.com/plotly/plotly.js/pull/3993/commits/1add90efdfb7d930ff3c0c880199cdeef776b5e2